third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TensorMemoryToLLVM.cpp

```diff
@@ -38,143 +91,170 @@ SmallVector<Value> packToI32(const SmallVector<Value> &values, Location loc,
+  numRegsPerMsg = std::min(numRegsPerMsg, numRegs);
+  // Invert the above formulas to calculate the effective runtime message width
+  colsPerMessage = (numRegsPerMsg * (TMemMsgT::opBitWidth / 32)) /
+                   (TMemMsgT::colsPerThread * TMemMsgT::rowsPerThread);
+
+  // Half as many registers are needed for 16-bit packed elements,
+  // so twice as many columns are accessed per message.
+  colsPerMessage *= info.numElementsPer32B;
+  for (int block = 0; block < info.numBlocks; block += info.numWarpGroups) {
     Value address = b.ptrtoint(i32_ty, baseAddress);
     Value blockId =
         b.add(b.i32_val(block),
-              b.udiv(warpGroupId, b.i32_val(numWarpGroupsPerBlock)));
-    Value blockRowId = rowId;
+              b.udiv(warpGroupId, b.i32_val(info.numWarpGroupsPerBlock)));
     Value warpGroupIdInBlock =
-        b.urem(warpGroupId, b.i32_val(numWarpGroupsPerBlock));
+        b.urem(warpGroupId, b.i32_val(info.numWarpGroupsPerBlock));
     Value startColumnId =
-        b.mul(warpGroupIdInBlock, b.i32_val(colsPerWarpGroup));
-    if (blocksInterleaved) {
+        b.mul(warpGroupIdInBlock, b.i32_val(info.colsPerWarpGroup));
+    Value blockRowId =
+        b.mul(warpIdInGroup, b.i32_val(TMemRuntimeInfo::numRowsPerWarp));
+
+    if (info.blocksInterleaved) {
       Value blockIdIsOdd = b.urem(blockId, b.i32_val(2));
       Value blockIdPrevEven = b.sub(blockId, blockIdIsOdd);
       blockRowId = b.add(blockRowId, b.mul(blockIdIsOdd, b.i32_val(16)));
       startColumnId =
           b.add(startColumnId,
-                b.mul(blockIdPrevEven, b.i32_val(numColsPerBlock / 2)));
+                b.mul(blockIdPrevEven, b.i32_val(info.numColsPerBlock / 2)));
     } else {
       startColumnId =
-          b.add(startColumnId, b.mul(blockId, b.i32_val(numColsPerBlock)));
+          b.add(startColumnId, b.mul(blockId, b.i32_val(info.numColsPerBlock)));
     }
-    address =
-        b.add(b.add(address, b.shl(blockRowId, b.i32_val(16))), startColumnId);
-
-    for (int colStart = 0; colStart < numColsPerBlock;
-         colStart += numColsPerMessage) {
-      Value startAddress = b.add(address, b.i32_val(colStart));
-
-      // Column offset of second half of the message in case of 16x32bx2
-      // message.
-      int secondHalfColOffset = useStridedMessage ? colsPerWarpGroup / 2 : 0;
-      createMemoryOp(startAddress, secondHalfColOffset, unpackedb16,
-                     numRegsPerMessage, useStridedMessage);
+
+    // A strided message accesses twice as many columns per message,
+    // thus half as many messages are required
+    int numColumns = info.useStridedMessage ? info.numColsPerBlock / 2
+                                            : info.numColsPerBlock;
+    for (int colStart = 0; colStart < numColumns; colStart += colsPerMessage) {
+      // For messages that span only 16 rows (e.g. 16x256b), multiple messages
+      // are required to cover the entire set of rows per warp.
```

Comment:
any chance we can simplify the code as we don't support those at the moment?

Reply:
It's only removing the for-loop and the statement below. For symmetry on the tmem access geometry I personally would prefer to keep it, I found it confusing initially why only a loop over columns was present. If you feel strongly, I can remove it until